ESO:418 updated the enhancement proposal to include the network policy for proxy (#1998)
openshift/enhancements: enhancements/external-secrets-operator/external-secrets-network-policy.md

api-approvers:
  - “@tgeer”
creation-date: 2025-09-12
last-updated:  2026-05-04
tracking-link: 
  - https://issues.redhat.com/browse/ESO-165
  - https://issues.redhat.com/browse/ESO-70

>   - https://redhat.atlassian.net/browse/RFE-8516

Reviewer: nit: Why not https://issues.redhat.com/browse/ESO-418?

Author: Done

- As a security engineer, I need to verify that all `external-secrets` pods have a default-deny policy and only allow traffic that is explicitly required for their function.
- As a `external-secrets` user, I need assurance that applying security policies will not break core functionalities like secret management or webhook validation.
- As an administrator, I want to configure and manage egress rules for `external-secrets` operands via the operator API or CRDs, so I can control which external services they are allowed to access.
- As an administrator on a proxy-configured cluster, I want the operator to automatically allow ESO pods to reach the proxy server, so I do not have to manually create egress NetworkPolicies after every install or upgrade.

> - As an administrator who manages proxy egress manually, I want to set `spec.appConfig.proxy.networkPolicyAllowProxyEgressAll: Unmanaged` to disable the operator's automatic proxy egress policy, so I retain full control over proxy traffic rules.

Reviewer: Any reason why we want name field with the desired effect? The desired effect should be the value instead.
```suggestion
- As an administrator who manages proxy egress manually, I want to set `spec.appConfig.proxy.networkPolicy: Unmanaged` to disable the operator's automatic proxy egress policy, so I retain full control over proxy traffic rules.
```

Author: I thought we can have some different name to avoid the confusion,as we already have a field called NetworkPolicies []NetworkPolicy  in the API section. 
If you suggest to change , i am fine with that also.

Reviewer: Since it's already listed under Proxy, I feel it does give out that meaning that NP is for proxy and how it should be handled. If you feel it could cause confusion, how about `NetworkPolicyProvisioning: [Managed, Unmanaged]`?

Author: okay as its under the proxy section have updated it with `networkPolicy`

Reviewer: On second thought, `NetworkPolicyProvisioning` seems more intuitive.

> - As a developer after upgrading from an older release, I want the operator to clean up legacy unprefixed NetworkPolicy objects after it has applied the new prefixed ones, so there are no stale or redundant policies left behind.

Reviewer 2: should we specify the release explicitly to make it clear, and perhaps prefix with something like [1.2.0] addition to make it clear for a first time reader of the EP?

Author: Done

> - As a cluster administrator upgrading from an older release, I want the operator to clean up legacy unprefixed NetworkPolicy objects after it has applied the new prefixed ones, so there are no stale or duplicate policies left behind.

Reviewer: ```suggestion
- As a developer after upgrading from an older release, I want the operator to clean up legacy unprefixed NetworkPolicy objects after it has applied the new prefixed ones, so there are no stale or redundant policies left behind.
```

Author: Done

- Define specific ingress and egress rules for `external-secrets` operand  to allow them to function correctly while blocking unnecessary traffic.
- Ensure that metrics collection for all components remains functional.
- Ensure the API server can communicate with the `webhook` for admission control.
- Automatically create a proxy-egress allow policy when an effective proxy is configured, controlled by `spec.appConfig.proxy.networkPolicyAllowProxyEgressAll` (default `Managed`).

> - Introduce a stable `eso-sys-` / `eso-user-` naming scheme for all operator-managed NetworkPolicy objects to enable unambiguous ownership and safe pruning.

Reviewer: ```suggestion
- Introduce a stable naming scheme (eso-sys- / eso-user-) for all operator-managed NetworkPolicy objects. This ensures unambiguous ownership and prevents naming overlaps with default policies, which currently lead to reconciliation conflicts
```

Author: Done

> - Migrate legacy unprefixed NetworkPolicy names to the new scheme on upgrade and clean up stale objects without leaving gaps in coverage.

Reviewer: ```suggestion
- Handle the migration of legacy, unprefixed NetworkPolicies to the new naming scheme during upgrades. This includes cleaning up stale objects while ensuring no gaps in traffic coverage during the transition.
```

Author: Done

### Non-Goals

- This enhancement does not propose creating a generic, cluster-wide policy management solution. The policies are specific to `external-secrets`.
- We are not introducing AdminNetworkPolicy at this stage, as standard NetworkPolicy objects are sufficient for this scope and can be managed directly by the operator.

> - Creating a user-facing option to disable these policies is not in scope, as they represent a baseline security posture.

Reviewer: Any reason why this was removed?

Author: because we can disable the proxy Np through the CR , hence removed it.

Reviewer: We are not disabling the proxy NP, we are giving a choice whether user want to configure it or operator should. This is done because, the traffic is to an external server, and other default NPs we create are internal components, where we are able to define a better policy, if not stringent.

Reviewer: Please include this on-goal again, if no other objection.

> - `spec.appConfig.proxy.networkPolicyAllowProxyEgressAll` does not affect non-proxy clusters — `getProxyConfiguration()` returns nil there and no proxy policy is ever created regardless of the field value.

Reviewer: Why is this a non-goal?

Author: This can be removed from non-goals

- Define specific ingress and egress rules for `external-secrets` operand  to allow them to function correctly while blocking unnecessary traffic.
- Ensure that metrics collection for all components remains functional.
- Ensure the API server can communicate with the `webhook` for admission control.
- Automatically create a proxy-egress allow policy when an effective proxy is configured, controlled by `spec.appConfig.proxy.networkPolicyProvisioning` (default `Managed`). When set to `Unmanaged`, the user is responsible for managing the proxy egress NetworkPolicy, giving a choice whether the operator or the user configures it.
- Introduce a stable naming scheme (`eso-sys-` / `eso-user-`) for all operator-managed NetworkPolicy objects. This ensures unambiguous ownership and prevents naming overlaps with default policies, which currently lead to reconciliation conflicts

> - Handle the migration of legacy, unprefixed NetworkPolicies to the new naming scheme during upgrades. This includes cleaning up stale objects while ensuring no gaps in traffic coverage during the transition.

Reviewer 2: is legacy the correct word? Or should be say versions prior to 1.2.0 to be more clear for a first time reader of the EP?

Author: true, updated the EP

        * **Allow Ingress from Core Controller:** Permit ingress from the `external-secrets` controller pod for communication with the Bitwarden server.
        * Note: The Bitwarden server is an optional integration. It is only created if explicitly enabled by user configuration.* User has to additionally create a allow NetworkPolicy to interact with the external `Bitwarden Secret Manager`.

4.  **Proxy Egress Policy (conditional):** After applying the static operand policies, the reconciler evaluates whether an automatic proxy egress allow policy should be created in the `external-secrets` namespace. The policy is created only when **both** conditions hold:

    * `spec.appConfig.proxy.networkPolicyAllowProxyEgressAll` is `Managed` (the default).

>     * * Proxy is configured in `ExternalSecretsManager` or `ExternalSecretsConfig` or in the OpenShift cluster.

Reviewer: nit
```suggestion
    * Proxy is configured in `ExternalSecretsManager` or `ExternalSecretsConfig` or in the OpenShift cluster.
```

Author: Done

    When created, this policy allows all ESO pods to reach the proxy server on its configured port. When either condition is not met (`Unmanaged`, or no proxy configured), the policy is not created; if one was previously created, it is removed on the next reconcile.

5.  **Naming scheme:** All operator-owned Kubernetes NetworkPolicy objects use a stable name prefix so the operator can unambiguously identify and prune them:

    * Static system policies (bindata): `eso-sys-<stable-name>` (e.g., `eso-sys-deny-all-traffic`).
    * Programmatically-built system policies : also use the `eso-sys-<stable-name>` prefix (e.g., `eso-sys-proxy-egress-core`). These are assembled in Go at reconcile time because they embed runtime data (e.g., the proxy port from `getProxyConfiguration()`).
    * User-defined policies from `spec.networkPolicies[]`: the operator prepends `eso-user-` to the logical name in the CR. The user writes `name: allow-external-secrets-egress`; the resulting Kubernetes object is named `eso-user-allow-external-secrets-egress`.

> 6.  **Migration and prune:** `cleanupMigratedNetworkPolicies()` runs only after the apply step fully succeeds. On the first reconcile after upgrade (no `migration-complete` annotation on the CR), it lists all NetworkPolicies in the namespace by label (`app.kubernetes.io/managed-by` + `app.kubernetes.io/part-of`) and by name, diffs against the already-applied desired set, and deletes any whose name is not in the desired set — this covers legacy unprefixed names as well as any stale user NPs. Once all deletions succeed, a `migration-complete` annotation is written to the `ExternalSecretsConfig` CR using the `fieldOwner` option. On subsequent reconciles the full deletion loop is skipped, but the label-based list and diff still runs every reconcile to catch newly stale entries (e.g. a user removes an NP from the CR). This function will be removed after 3 releases, by which point every cluster will have been reconciled at least once under the new naming scheme.

Reviewer: Migration is broader and I think it could mislead users when the annotation is noticed.

Author: updated with `externalsecretsconfig.operator.openshift.io/skip-np-cleanup-check: true`

6.  **Migration and prune:** `cleanupMigratedNetworkPolicies()` runs only after the apply step fully succeeds. On each reconcile it operates as follows:
    1. Check whether the annotation `externalsecretsconfig.operator.openshift.io/skip-np-cleanup-check: "true"` is present on the `ExternalSecretsConfig` CR. If the annotation is present, skip the full deletion loop entirely.
    2. If the annotation is absent, list all NetworkPolicy resources in the namespace by label (`app.kubernetes.io/managed-by: external-secrets-operator` and `app.kubernetes.io/part-of: external-secrets-operator`) and by name, and delete any whose name is not in the desired set. The legacy unprefixed policies — `deny-all-traffic`, `allow-to-dns`, `allow-api-server-egress-for-main-controller`, etc. — carry the operator labels but are no longer in the desired set, so they get cleaned up. Stale user NPs are also removed by the same diff to avoid duplicate NPs.

>     3. After all deletions succeed, write the `externalsecretsconfig.operator.openshift.io/skip-np-cleanup-check: "true"` annotation onto the `ExternalSecretsConfig` CR using the `fieldOwner` option so subsequent reconciles skip the deletion loop and avoid unnecessary GETs.On subsequent reconciles (annotation present), the full deletion loop is skipped, but the label-based list and diff still runs every reconcile to catch newly stale entries (e.g. a user removes an NP from the CR).

Reviewer: This is not true correct? Currently we don't allow removing policies, so do we really do perform this?
```
On subsequent reconciles (annotation present), the full deletion loop is skipped, but the label-based list and diff still runs every reconcile to catch newly stale entries (e.g. a user removes an NP from the CR).
```

Author: True, updated

        * **Allow Ingress from Core Controller:** Permit ingress from the `external-secrets` controller pod for communication with the Bitwarden server.
        * Note: The Bitwarden server is an optional integration. It is only created if explicitly enabled by user configuration.* User has to additionally create a allow NetworkPolicy to interact with the external `Bitwarden Secret Manager`.

4.  **Proxy Egress Policy (conditional):** After applying the static operand policies, the reconciler evaluates whether an automatic proxy egress allow policy should be created in the `external-secrets` namespace. The policy is created only when **both** conditions hold:

    * `spec.appConfig.proxy.networkPolicyAllowProxyEgressAll` is `Managed` (the default).

>     * `getProxyConfiguration()` returns a non-nil result - i.e., an effective proxy is actually configured on the cluster.

Reviewer: Please rephrase accordingly
```suggestion
    * Proxy is configured in ExternalSecretsManager or ExternalSecretsConfig or in the OpenShift cluster.
```

Author: Done

5.  **Naming scheme:** All operator-owned Kubernetes NetworkPolicy objects use a stable name prefix so the operator can unambiguously identify and prune them:

    * Static system policies (bindata): `eso-sys-<stable-name>` (e.g., `eso-sys-deny-all-traffic`).
    * Programmatically-built system policies : also use the `eso-sys-<stable-name>` prefix (e.g., `eso-sys-proxy-egress-core`). These are assembled in Go at reconcile time because they embed runtime data (e.g., the proxy port from `getProxyConfiguration()`).
    * User-defined policies from `spec.networkPolicies[]`: the operator prepends `eso-user-` to the logical name in the CR. The user writes `name: allow-external-secrets-egress`; the resulting Kubernetes object is named `eso-user-allow-external-secrets-egress`.

6.  **Migration and prune:** `cleanupMigratedNetworkPolicies()` runs only after the apply step fully succeeds. On each reconcile it operates as follows:
    1. Check whether the annotation `externalsecretsconfig.operator.openshift.io/skip-np-cleanup-check: "true"` is present on the `ExternalSecretsConfig` CR. If the annotation is present, skip the full deletion loop entirely.

>     2. If the annotation is absent, list all NetworkPolicy resources in the namespace by label (`app.kubernetes.io/managed-by: external-secrets-operator` and `app.kubernetes.io/part-of: external-secrets-operator`) and by name, and delete any whose name is not in the desired set. The legacy unprefixed policies — `deny-all-traffic`, `allow-to-dns`, `allow-api-server-egress-for-main-controller`, etc. — carry the operator labels but are no longer in the desired set, so they get cleaned up. Stale user NPs are also removed by the same diff to avoid duplicate NPs.

Reviewer 2: ```suggestion
    2. If the annotation is absent, list all NetworkPolicy resources in the namespace by label (`app.kubernetes.io/managed-by: external-secrets-operator` and `app.kubernetes.io/part-of: external-secrets-operator`) by name, and delete any whose name does not begin with prefix eso-user or eso-system. The legacy unprefixed policies — `deny-all-traffic`, `allow-to-dns`, `allow-api-server-egress-for-main-controller`, etc. which carry the operator labels get cleaned up.
```

Author: Done

          - {} # Allow all egress for external issuers communication
    ```  

7. **Proxy Egress Policy (auto-created, conditional):** This policy is built **programmatically in the reconciler** - it is not a static bindata manifest because the proxy port is runtime data obtained from `getProxyConfiguration()`.

    The policy is applied only when **both** conditions hold:

    * `spec.appConfig.proxy.networkPolicyAllowProxyEgressAll` is `Managed` (the default).
    * `getProxyConfiguration()` returns a non-nil result - i.e., an effective proxy is actually configured on the cluster.

    If either condition is not met, this policy is not created (or removed if previously present).

    Illustrative structure (port value filled in at runtime):

    ```yaml
    apiVersion: networking.k8s.io/v1
    kind: NetworkPolicy
    metadata:
      name: eso-sys-proxy-egress-core
      namespace: external-secrets
      labels:
        app.kubernetes.io/managed-by: external-secrets-operator
    spec:
      podSelector:
        matchLabels:
          app.kubernetes.io/name: external-secrets
      policyTypes:
        - Egress
      egress:
        - ports:
            - protocol: TCP

>               port: <proxy-port>   # set at reconcile time from getProxyConfiguration()

Reviewer: The EP should also talk about, how the port will be arrived at like when an URL is provided without port, and arriving at 443 as default for https and 80 as default for http.

Author: Added a section for this

            - protocol: TCP
              port: 6443
    ```  
6. **User-Configurable Policies:** Users configure additional policies via the `ExternalSecretsConfig` custom resource to set `external-secrets` controller egress allow policy to communicate with external providers. The operator prepends `eso-user-` to the logical name in the CR. The user writes `name: allow-external-secrets-egress`; the resulting Kubernetes object is named `eso-user-allow-external-secrets-egress`.
Example user configuration:

    ```yaml
    apiVersion: operator.openshift.io/v1alpha1
    kind: ExternalSecretsConfig
    metadata:
      name: cluster
    spec:
      appConfig:
        proxy:
          networkPolicyProvisioning: Managed   # default; set Unmanaged to manage proxy egress yourself
      networkPolicies:
        - name: allow-external-secrets-egress    # K8s object: eso-user-allow-external-secrets-egress
          componentName: CoreController
          policyTypes:
          - Egress
          egress:
          - {} # Allow all egress for external issuers communication
    ```  

> 7. **Proxy Egress Policy (auto-created, conditional):** This policy is built **programmatically in the reconciler** - it is not a static bindata manifest because the proxy port is runtime data obtained from `getProxyConfiguration()`.

Reviewer 2: do we need to specify "static _bindata_ manifest" in the EP or should "static manifest" be good enough if we want to restructure the implementation later as to where the static manifests should be stored?

Author: Done

        // +optional
        // +listType=atomic
        Egress []networkingv1.NetworkPolicyEgressRule `json:"egress,omitempty" protobuf:"bytes,3,rep,name=egress"`
  }

    // NetworkPolicyEgressManagement controls whether the operator manages the proxy egress NetworkPolicy.

>     type NetworkPolicyEgressManagement string

Reviewer: ```suggestion
    // ManagementState controls whether the operator manages the resource lifecycle.
    type ManagementState string
```

Author: Done

    const (
        // NetworkPolicyEgressManaged means the operator automatically creates and reconciles
        // the eso-sys-proxy-egress-core NetworkPolicy when a proxy is configured.

>         NetworkPolicyEgressManaged NetworkPolicyEgressManagement = "Managed"
> 
>         // NetworkPolicyEgressUnmanaged means the user is responsible for proxy egress;
>         // the operator does not create, update, or delete the proxy egress NetworkPolicy.
>         NetworkPolicyEgressUnmanaged NetworkPolicyEgressManagement = "Unmanaged"
>     )

Reviewer: ```suggestion
const (
    // Managed indicates the Operator is responsible for the resource lifecycle.
    ManagementStateManaged   ManagementState = "Managed"
    // Unmanaged indicates the User is responsible for the resource lifecycle.
    ManagementStateUnmanaged ManagementState = "Unmanaged"
)
```

Author: Done

>     // ProxyConfig is extended to include the proxy egress NetworkPolicy management field.

Reviewer: ```suggestion
```

Author: Done

    type ProxyConfig struct {
        // ... existing httpProxy, httpsProxy, noProxy fields ...

        // networkPolicy controls whether the operator automatically
        // creates and manages the "eso-sys-proxy-egress-core" NetworkPolicy that allows
        // all ESO pods to reach the cluster proxy server.
        // On clusters without a proxy configured, this field has no effect regardless of value.
        // +kubebuilder:validation:Enum=Managed;Unmanaged
        // +kubebuilder:default=Managed
        // +optional

>         NetworkPolicy NetworkPolicyEgressManagement `json:"networkPolicy,omitempty"`

Reviewer: ```suggestion
        // NetworkPolicy defines the management strategy for the proxy egress rule.
        // When set to Managed, the operator automatically provisions and maintains
        // a NetworkPolicy allowing traffic to the configured proxy.
        // If no proxy is configured, no NetworkPolicy will be created
        // regardless of this setting.
        // +kubebuilder:validation:Enum=Managed;Unmanaged
        // +kubebuilder:default=Managed
        // +optional
        NetworkPolicy ManagementState `json:"networkPolicy,omitempty"`
```

Author: Done

## Upgrade / Downgrade Strategy

* **Upgrade:** On upgrade, the operator applies the new `eso-sys-` prefixed NetworkPolicy objects first, then runs `cleanupMigratedNetworkPolicies()` to remove unprefixed objects from operator versions prior to 1.2.0 (e.g., `deny-all-traffic`, `allow-to-dns`, etc.). This apply-before-delete ordering ensures there is no window where a required policy is absent.

  This function runs only after the apply step fully succeeds. On each reconcile it operates as follows:
  1. Check whether the annotation `externalsecretsconfig.operator.openshift.io/skip-np-cleanup-check: "true"` is present on the `ExternalSecretsConfig` CR. If the annotation is present, skip the full deletion loop entirely.

>   2. If the annotation is absent, list all NetworkPolicy resources in the namespace by label (`app.kubernetes.io/managed-by: external-secrets-operator` and `app.kubernetes.io/part-of: external-secrets-operator`) and by name, and delete any whose name is not in the desired set. The unprefixed policies from operator versions prior to 1.2.0 - `deny-all-traffic`, `allow-to-dns`, `allow-api-server-egress-for-main-controller`, etc. - carry the operator labels but are no longer in the desired set, so they get cleaned up. Stale user NPs are also removed by the same diff to avoid the duplicate NPs.

Reviewer 2: This is verbatim repeating from chapter 6. migration and pruning in workflow description.
Can we make a separate section and refer to it from both workflow description and the current chapter?
